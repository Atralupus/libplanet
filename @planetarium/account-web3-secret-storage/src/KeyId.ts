import { crypto } from "#crypto";

export type KeyId = string;

const UUID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function isKeyId(keyId: string): keyId is KeyId {
  return !!keyId.match(UUID_PATTERN);
}

export function generateKeyId(): KeyId {
  return crypto.randomUUID();
}

export default KeyId;
