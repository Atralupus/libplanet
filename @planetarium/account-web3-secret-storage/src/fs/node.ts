import * as fs from "node:fs/promises";

export async function mkdir(
  path: string,
  options?: { recursive: true },
): Promise<void> {
  await fs.mkdir(path, options);
}

export function readFile(
  path: string,
  options?: { encoding: "utf8" },
): Promise<string> {
  return fs.readFile(path, options) as Promise<string>;
}

export function removeFile(path: string): Promise<void> {
  return fs.unlink(path);
}

export async function* listFiles(directory: string): AsyncIterable<string> {
  let dir;
  try {
    dir = await fs.opendir(directory);
  } catch (e) {
    if (
      typeof e === "object" &&
      e != null &&
      "code" in e &&
      e.code === "ENOENT"
    ) {
      // In case where there is no directory at all (it's likely the first
      // time to run this operation in a system), it should be considered
      // it's just empty (instead of considering it an exceptional case).
      return;
    }
    throw e;
  }
  for await (const dirEntry of dir) {
    if (!dirEntry.isFile()) continue;
    yield dirEntry.name;
  }
}

export function writeFile(
  path: string,
  content: string,
  encoding?: "utf8",
): Promise<void> {
  return fs.writeFile(path, content, encoding);
}
