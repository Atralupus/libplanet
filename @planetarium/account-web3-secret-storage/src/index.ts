export { KeyId } from "./KeyId.js";
export { PassphraseEntry } from "./PassphraseEntry.js";
export { TtyPassphraseEntry } from "./TtyPassphraseEntry.js";
export {
  type DecryptionOptions,
  WeakPrivateKeyError,
  Web3Account,
  type Web3KeyObject,
} from "./Web3Account.js";
export { getDefaultWeb3KeyStorePath, Web3KeyStore } from "./Web3KeyStore.js";
