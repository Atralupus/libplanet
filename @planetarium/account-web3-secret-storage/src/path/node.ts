export { join } from "node:path";

/**
 * Determines the default key store path.  It depends on the platform:
 *
 * - Linux/macOS: `$HOME/.config/planetarium/keystore`
 * - Windows: `%AppData%\planetarium\keystore`
 */
export function getDefaultWeb3KeyStorePath(): string {
  const { homedir } = require("node:os");
  const path = require("node:path");
  const baseDir =
    process.platform === "win32"
      ? process.env.AppData || path.join(homedir(), "AppData", "Roaming")
      : process.env.XDG_CONFIG_HOME || path.join(homedir(), ".config");
  // Note that it's not necessary to explicitly choose one of `path.win32` or
  // `path.posix` here, but it makes unit tests less dependent on mocks:
  return (process.platform === "win32" ? path.win32 : path.posix).join(
    baseDir,
    "planetarium",
    "keystore",
  );
}
