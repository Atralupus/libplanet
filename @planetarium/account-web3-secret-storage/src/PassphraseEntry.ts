import { KeyId } from "./KeyId.js";

export type Passphrase = Uint8Array | string;

export interface PassphraseEntry {
  authenticate(keyId: KeyId, firstAttempt: boolean): Promise<Passphrase>;
  configurePassphrase(): Promise<Passphrase>;
}
