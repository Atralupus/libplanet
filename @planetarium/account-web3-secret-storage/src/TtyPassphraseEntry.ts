import { KeyId } from "./KeyId.js";
import { Passphrase, PassphraseEntry } from "./PassphraseEntry.js";
import * as readline from "node:readline";
import { Readable, Writable } from "node:stream";

export interface PassphraseReaderOptions {
  input: Readable;
  output: Writable;
  mask: string | false;
}

function isNewLine(chunk: Buffer | string) {
  return (
    (typeof chunk === "string" && chunk.endsWith("\n")) ||
    (chunk instanceof Buffer && chunk.at(chunk.length - 1) === 0x0a)
  );
}

export function readPassphrase(
  prompt: string,
  { input, output, mask }: PassphraseReaderOptions,
): Promise<Passphrase> {
  let masking = false;
  const rl = readline.createInterface({
    input,
    output: new Writable({
      write: (chunk, encoding: BufferEncoding, callback) => {
        if (masking && !isNewLine(chunk)) {
          output.write(
            mask,
            encoding.endsWith("buffer") ? "utf8" : encoding,
            callback,
          );
        } else output.write(chunk, encoding, callback);
      },
    }),
    terminal: true,
  });

  return new Promise((resolve) => {
    rl.question(prompt, (answer) => {
      rl.close();
      resolve(answer);
    });
    setTimeout(() => (masking = mask !== false), 0); // wait for I/O interrupt
  });
}

const prompts: {
  authenticate: Record<string, string>;
  authenticateRetry: Record<string, string>;
  configurePassphrase: Record<string, string>;
  configurePassphraseRetry: Record<string, string>;
  confirmConfigurePassphrase: Record<string, string>;
} = {
  authenticate: {
    "": "Passphrase for {keyId}: ",
    ko: "{keyId} 키의 암호: ",
    qts: "AUTHENTICATE:{keyId}", // for unit test
  },
  authenticateRetry: {
    "": "Incorrect; try again: ",
    ko: "틀림. 재시도: ",
    qts: "AUTHENTICATE_RETRY:{keyId}", // for unit test
  },
  configurePassphrase: {
    "": "New passphrase: ",
    ko: "새 암호: ",
    qts: "CONFIGURE_PASSPHRASE", // for unit test
  },
  configurePassphraseRetry: {
    "": "Mismatch; new passphrase: ",
    ko: "틀림. 새 암호 재입력:",
    qts: "CONFIGURE_PASSPHRASE_RETRY", // for unit test
  },
  confirmConfigurePassphrase: {
    "": "Confirm passphrase: ",
    ko: "암호 확인:",
    qts: "CONFIRM_CONFIGURE_PASSPHRASE", // for unit test
  },
};

function getLocale(): string {
  return Intl.DateTimeFormat().resolvedOptions().locale;
}

interface TtyPassphraseEntryOptions extends PassphraseReaderOptions {
  locale: string;
}

export class TtyPassphraseEntry implements PassphraseEntry {
  #options: TtyPassphraseEntryOptions;

  constructor(options: Partial<TtyPassphraseEntryOptions> = {}) {
    this.#options = {
      input: options.input ?? process.stdin,
      output: options.output ?? process.stdout,
      mask: options.mask ?? "*",
      locale: options.locale ?? getLocale(),
    };
  }

  #getPrompt(promptType: keyof typeof prompts) {
    const table = prompts[promptType];
    let locale = this.#options.locale;
    let msg;
    while (msg == null) {
      msg = table[locale];
      locale = locale.replace(/(^|[-_.])[^-_.]*$/, "");
    }
    return msg;
  }

  authenticate(keyId: KeyId, firstAttempt: boolean): Promise<Passphrase> {
    const message = this.#getPrompt(
      firstAttempt ? "authenticate" : "authenticateRetry",
    );
    const prompt = message.replace(/\{keyId\}/g, keyId);
    return readPassphrase(prompt, this.#options);
  }

  async configurePassphrase(): Promise<Passphrase> {
    let prompt = this.#getPrompt("configurePassphrase");
    const retryPrompt = this.#getPrompt("configurePassphraseRetry");
    const confirmPrompt = this.#getPrompt("confirmConfigurePassphrase");
    while (true) {
      const passphrase = await readPassphrase(prompt, this.#options);
      const confirmPassphrase = await readPassphrase(
        confirmPrompt,
        this.#options,
      );
      if (passphrase === confirmPassphrase) return passphrase;
      prompt = retryPrompt;
    }
  }
}
