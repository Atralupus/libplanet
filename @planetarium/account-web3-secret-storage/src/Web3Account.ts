import { KeyId } from "./KeyId.js";
import { PassphraseEntry } from "./PassphraseEntry.js";
import {
  ExportableAccount,
  PublicKey,
  RawPrivateKey,
  Signature,
} from "@planetarium/account";
import { type KeystoreAccount, decryptKeystoreJson } from "ethers";

export type Web3KeyObjectKdf =
  | {
      kdf: "pbkdf2";
      kdfparams: {
        c: number;
        dklen: number;
        prf: "hmac-sha256";
        salt: string;
      };
    }
  | {
      kdf: "scrypt";
      kdfparams: {
        dklen: number;
        n: number;
        p: number;
        r: number;
        salt: string;
      };
    };

export interface Web3KeyObject {
  version: 3;
  id: KeyId;
  address: string;
  crypto: {
    cipher: "aes-128-ctr" | "aes-128-cbc";
    cipherparams: {
      iv: string;
    };
    ciphertext: string;
    mac: string;
  } & Web3KeyObjectKdf;
}

export class Web3Account implements ExportableAccount {
  #keyObject: Web3KeyObject;
  #passphraseEntry: PassphraseEntry;

  constructor(keyObject: Web3KeyObject, passphraseEntry: PassphraseEntry) {
    this.#keyObject = keyObject;
    this.#passphraseEntry = passphraseEntry;
  }

  async exportPrivateKey(): Promise<RawPrivateKey> {
    let firstAttempt = true;
    let account: KeystoreAccount;
    while (true) {
      const passphrase = await this.#passphraseEntry.authenticate(
        this.#keyObject.id,
        firstAttempt,
      );
      let currentProgress = 0;
      const json = JSON.stringify(this.#keyObject);
      try {
        // TODO: ethers does not support scrypt yet, however Libplanet.KeyStore.Web3KeyStore does.
        // We need to support scrypt in the future.
        account = await decryptKeystoreJson(json, passphrase, (progress) => {
          currentProgress = progress;
        });
      } catch (e) {
        if (currentProgress <= 0) {
          throw e;
        }
        firstAttempt = false;
        continue;
      }
      break;
    }
    return RawPrivateKey.fromHex(account.privateKey.replace(/^0x/i, ""));
  }

  async getPublicKey(): Promise<PublicKey> {
    const key = await this.exportPrivateKey();
    return await key.getPublicKey();
  }

  async sign(message: Uint8Array): Promise<Signature> {
    const key = await this.exportPrivateKey();
    return await key.sign(message);
  }
}
