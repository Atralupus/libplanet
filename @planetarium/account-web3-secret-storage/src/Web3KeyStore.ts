import { generateKeyId, KeyId } from "./KeyId.js";
import { PassphraseEntry } from "./PassphraseEntry.js";
import {
  encryptKeyObject,
  isKeyObject,
  type Web3AccountOptions,
  Web3Account,
  Web3KeyObject,
} from "./Web3Account.js";
import {
  type AccountDeletion,
  type AccountGeneration,
  type AccountImportation,
  type AccountMetadata,
  type AccountRetrieval,
  Address,
  type ImportableKeyStore,
  RawPrivateKey,
} from "@planetarium/account";
import { join, getDefaultWeb3KeyStorePath } from "#path";
export { getDefaultWeb3KeyStorePath } from "#path";
import * as fs from "#fs";

export interface Web3KeyStoreOptions {
  path?: string;
  passphraseEntry: PassphraseEntry;
}

const pattern =
  /^(?:UTC--([0-9]{4}-[0-9]{2}-[0-9]{2})T([0-9]{2}-[0-9]{2}-[0-9]{2})Z--)?([0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12})(?:.json)?$$/i;
export function parseKeyFilename(
  name: string,
): undefined | { keyId: KeyId; createdAt?: Date } {
  const match = pattern.exec(name);
  if (match == null) return undefined;
  return {
    keyId: match[3].toLowerCase(),
    createdAt:
      match[1] != null && match[2] != null
        ? new Date(`${match[1]}T${match[2].replace(/-/g, ":")}Z`)
        : undefined,
  };
}

export interface Web3KeyMetadata {
  address?: Address;
}

export class Web3KeyStore
  implements ImportableKeyStore<KeyId, Web3Account, Web3KeyMetadata>
{
  readonly #passphraseEntry: PassphraseEntry;
  readonly #accountOptions: Partial<Web3AccountOptions>;

  readonly path: string;

  constructor(options: Web3KeyStoreOptions & Partial<Web3AccountOptions>) {
    this.path = options.path ?? getDefaultWeb3KeyStorePath();
    this.#passphraseEntry = options.passphraseEntry;
    this.#accountOptions = options;
  }

  async *#listKeyFiles(): AsyncIterable<string> {
    for await (const name of fs.listFiles(this.path)) {
      yield name;
    }
  }

  async #getKeyPath(
    keyId: KeyId,
  ): Promise<{ path: string; keyId: string; createdAt?: Date } | undefined> {
    for await (const name of this.#listKeyFiles()) {
      const parsed = parseKeyFilename(name);
      if (parsed != null && parsed.keyId === keyId) {
        return { ...parsed, path: join(this.path, name) };
      }
    }
    return undefined;
  }

  async *list(): AsyncIterable<AccountMetadata<KeyId, Web3KeyMetadata>> {
    for await (const name of this.#listKeyFiles()) {
      const parsed = parseKeyFilename(name);
      if (parsed == null) continue;
      const keyPath = join(this.path, name);
      let json: unknown;
      try {
        json = JSON.parse(await fs.readFile(keyPath, { encoding: "utf8" }));
      } catch (_) {
        continue;
      }
      if (!isKeyObject(json)) continue;
      let address: Address;
      try {
        address = Address.fromHex(json.address, true);
      } catch (_) {
        continue;
      }
      const metadata: Web3KeyMetadata = { address };
      yield { ...parsed, metadata };
    }
  }

  async get(
    keyId: Readonly<KeyId>,
  ): Promise<AccountRetrieval<KeyId, Web3Account, Web3KeyMetadata>> {
    const keyPath = await this.#getKeyPath(keyId);
    if (keyPath == null) return { result: "keyNotFound", keyId };
    let json;
    try {
      json = await fs.readFile(keyPath.path, {
        encoding: "utf8",
      });
    } catch (e) {
      if (
        e != null &&
        typeof e === "object" &&
        "code" in e &&
        e.code === "ENOENT"
      ) {
        return { result: "keyNotFound", keyId };
      }
      return { result: "error", keyId, message: `${e}` };
    }
    const keyObject: unknown = JSON.parse(json);
    if (!isKeyObject(keyObject)) {
      return { result: "error", keyId, message: "Invalid key file" };
    }
    const account = new Web3Account(
      keyObject,
      this.#passphraseEntry,
      this.#accountOptions,
    );
    return {
      result: "success",
      account,
      keyId,
      metadata: { address: await account.getAddress() },
      createdAt: keyPath.createdAt,
    };
  }

  async generate(
    metadata?: Partial<Web3KeyMetadata>,
  ): Promise<AccountGeneration<KeyId, Web3Account>> {
    if (metadata?.address != null) {
      return {
        result: "error",
        message: "Address cannot be predetermined before generating key",
      };
    }
    const privateKey = await RawPrivateKey.generate();
    const result = await this.#import(privateKey, metadata);
    if (result.result === "success") {
      return {
        result: "success",
        keyId: result.keyId,
        account: new Web3Account(result.keyObject, this.#passphraseEntry),
      };
    }
    return result;
  }

  async delete(keyId: Readonly<KeyId>): Promise<AccountDeletion<KeyId>> {
    const keyPath = await this.#getKeyPath(keyId);
    if (keyPath == null) return { result: "keyNotFound", keyId };
    try {
      await fs.removeFile(keyPath.path);
    } catch (e) {
      return { result: "error", message: `${e}` };
    }
    return { result: "success", keyId };
  }

  async #import(
    privateKey: RawPrivateKey,
    metadata?: Partial<Web3KeyMetadata>,
  ): Promise<
    | {
        readonly result: "success";
        readonly keyId: KeyId;
        readonly keyObject: Web3KeyObject;
      }
    | {
        readonly result: "error";
        readonly message?: string;
      }
  > {
    if (
      metadata?.address != null &&
      !metadata.address.equals(await Address.deriveFrom(privateKey))
    ) {
      return {
        result: "error",
        message:
          "Address does not match the private key " +
          "(hint: you do not have to specify it manually)",
      };
    }
    const passphrase = await this.#passphraseEntry.configurePassphrase();
    const keyId = generateKeyId();
    const keyObject = await encryptKeyObject(keyId, privateKey, passphrase);
    try {
      await fs.mkdir(this.path, { recursive: true });
    } catch (e) {
      return { result: "error", message: `${e}` };
    }
    const createdAt = new Date();
    const keyPath = join(
      this.path,
      `UTC--${createdAt
        .toISOString()
        .replace(/\.[0-9]+Z$/, "Z")
        .replace(/:/g, "-")}--${keyId}`,
    );
    try {
      await fs.writeFile(keyPath, JSON.stringify(keyObject), "utf8");
    } catch (e) {
      return { result: "error", message: `${e}` };
    }
    return { result: "success", keyId, keyObject };
  }

  async import(
    privateKey: RawPrivateKey,
    metadata?: Partial<Web3KeyMetadata>,
  ): Promise<AccountImportation<KeyId>> {
    const bytes = await privateKey.toBytes();
    if (bytes.at(0) === 0x00 && !this.#accountOptions.allowWeakPrivateKey) {
      return {
        result: "error",
        message:
          "The private key given is too weak; keys of length less than 32 bytes " +
          "are disallowed by default.  See also " +
          "the Web3AccountOptions.allowWeakPrivateKey option.",
      };
    }

    const result = await this.#import(privateKey, metadata);
    if (result.result === "success") {
      return { result: "success", keyId: result.keyId };
    }
    return result;
  }
}

export default Web3KeyStore;
