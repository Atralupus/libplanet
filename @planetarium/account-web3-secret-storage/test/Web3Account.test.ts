import { PublicKey, RawPrivateKey } from "@planetarium/account";
import * as fc from "fast-check";
import * as crypto from "node:crypto";
import { describe, expect, test } from "vitest";
import { KeyId } from "../src/KeyId";
import { Passphrase } from "../src/PassphraseEntry";
import {
  Web3AccountOptions,
  decryptKeyObject,
  encryptKeyObject,
  IncorrectPassphraseError,
  WeakPrivateKeyError,
  Web3Account,
  Web3KeyObject,
} from "../src/Web3Account";
import { MockPassphraseEntry } from "./MockPassphraseEntry";

const pbkdf2KeyObject: Web3KeyObject = {
  version: 3,
  id: "3c7bec5e-1f1d-4754-a1ce-3644ce1130f1",
  address: "98a253783288c3971cf7960157b8e053bd263da7",
  crypto: {
    ciphertext:
      "da63b632d3e48de4099e32e3664741fa60880547ab8740854abb77a8c5183638",
    cipherparams: {
      iv: "59b8523ebeaa3f37ea06898ff0860340",
    },
    cipher: "aes-128-ctr",
    kdfparams: {
      c: 10240,
      dklen: 32,
      prf: "hmac-sha256",
      salt: "debd6928cf96f23e3cd5580aebd911c8ea4eb7c29d4db10c251ce8f29c2f32f6",
    },
    kdf: "pbkdf2",
    mac: "5ea19f4c94faa5c698dbbcdd55390f72dac8b6383b2c1010328dfc9248eac2fc",
  },
};

const scryptKeyObject: Web3KeyObject = {
  crypto: {
    cipher: "aes-128-ctr",
    cipherparams: {
      iv: "83dbcc02d8ccb40e466191a123791e0e",
    },
    ciphertext:
      "d172bf743a674da9cdad04534d56926ef8358534d458fffccd4e6ad2fbde479c",
    kdf: "scrypt",
    kdfparams: {
      dklen: 32,
      n: 262144,
      p: 8,
      r: 1,
      salt: "ab0c7876052600dd703518d6fc3fe8984592145b591fc8fb5c6d43190334ba19",
    },
    mac: "2103ac29920d71da29f15d75b4a16dbe95cfd7ff8faea1056c33131d846e3097",
  },
  id: "3198bc9c-6672-5ab3-d995-4942343ae5b6",
  address: "008aeeda4d805471df9b2a5b0f38a0c3bcba786b",
  version: 3,
};

const insufficientLengthKeyObject: Web3KeyObject = {
  crypto: {
    cipher: "aes-128-ctr",
    cipherparams: {
      iv: "24ece9686a10f8687dc70ca42873a063",
    },
    ciphertext: "14",
    kdf: "pbkdf2",
    kdfparams: {
      c: 10240,
      dklen: 32,
      prf: "hmac-sha256",
      salt: "99b480642f6406a10c67429168fb36ba35da7743965e10390eda649ec6da2ae8",
    },
    mac: "7856d88908c72f0af78457ccd085696ad6193f7d2bda525723862f450ce060fa",
  },
  id: "b35a2647-8581-43ff-a98e-6083dc952632",
  address: "eb9afe072c781401bf364224c75a036e4d832f52",
  version: 3,
};

interface TestVector {
  keyObject: Web3KeyObject;
  passphrase: Passphrase;
  privateKey: RawPrivateKey;
  options?: Partial<Web3AccountOptions>;
}

const testVectors: Record<string, TestVector> = {
  pbkdf: {
    keyObject: pbkdf2KeyObject,
    passphrase: "passphrase",
    privateKey: RawPrivateKey.fromHex(
      "2af8da24b177f6723d8b99762b416fc3050289b572b8819ebbb0a6b143b1d55d",
    ),
  },
  scrypt: {
    keyObject: scryptKeyObject,
    passphrase: "testpassword",
    privateKey: RawPrivateKey.fromHex(
      "7a28b5ba57c53603b0b07b56bba752f7784bf506fa95edc395f5cf6c7514fe9d",
    ),
  },
  insufficientLength: {
    keyObject: insufficientLengthKeyObject,
    passphrase: "1",
    privateKey: RawPrivateKey.fromHex(
      "00000000000000000000000000000000000000000000000000000000000000aa",
    ),
    options: {
      allowWeakPrivateKey: true,
    },
  },
};

describe("Web3Account", () => {
  describe("getAddress()", () => {
    for (const [kdf, { keyObject, passphrase, options }] of Object.entries(
      testVectors,
    )) {
      test(kdf, async () => {
        const passphraseEntry = new MockPassphraseEntry(passphrase);
        const key = new Web3Account(keyObject, passphraseEntry, options);
        expect((await key.getAddress()).toHex("lower")).toBe(keyObject.address);
      });
    }
  });
  describe("getPublicKey()", () => {
    test("fastcheck", async () => {
      const passphraseEntry = new MockPassphraseEntry();
      const key = new Web3Account(pbkdf2KeyObject, passphraseEntry);

      expect(passphraseEntry.authenticateCalls).toStrictEqual([]);
      expect(passphraseEntry.configurePassphraseCalls).toBe(0);
      const publicKey = await key.getPublicKey();
      expect(passphraseEntry.authenticateCalls).toStrictEqual([
        { keyId: "3c7bec5e-1f1d-4754-a1ce-3644ce1130f1", firstAttempt: true },
      ]);
      expect(passphraseEntry.configurePassphraseCalls).toBe(0);

      expect(publicKey).toBeInstanceOf(PublicKey);
      expect(publicKey.toHex("compressed")).toStrictEqual(
        "03dfd25cea92d828d5c3b2836ffe5a843854285027b9de2dcca0920c0ff0beb5c4",
      );
    });

    for (const [
      kdf,
      { keyObject, passphrase, privateKey, options },
    ] of Object.entries(testVectors)) {
      test(kdf, async () => {
        const passphraseEntry = new MockPassphraseEntry(passphrase);
        const key = new Web3Account(keyObject, passphraseEntry, options);

        expect(passphraseEntry.authenticateCalls).toStrictEqual([]);
        expect(passphraseEntry.configurePassphraseCalls).toBe(0);
        const publicKey = await key.getPublicKey();
        expect(passphraseEntry.authenticateCalls).toStrictEqual([
          { keyId: keyObject.id, firstAttempt: true },
        ]);
        expect(passphraseEntry.configurePassphraseCalls).toBe(0);

        expect(publicKey).toBeInstanceOf(PublicKey);
        expect(publicKey.toHex("compressed")).toStrictEqual(
          await (await privateKey.getPublicKey()).toHex("compressed"),
        );

        if (options?.allowWeakPrivateKey) {
          const keyWithStricterOptions = new Web3Account(
            keyObject,
            passphraseEntry,
            {
              allowWeakPrivateKey: false,
            },
          );
          const publicKey2 = await key.getPublicKey();
          expect(publicKey2).toBeInstanceOf(PublicKey);
          expect(publicKey2.toHex("compressed")).toStrictEqual(
            await (await privateKey.getPublicKey()).toHex("compressed"),
          );
        }
      });
    }
  });

  describe("sign()", () => {
    test("fastcheck", async () => {
      const pubKey = PublicKey.fromHex(
        "03dfd25cea92d828d5c3b2836ffe5a843854285027b9de2dcca0920c0ff0beb5c4",
        "compressed",
      );
      await fc.assert(
        fc.asyncProperty(fc.uint8Array(), async (msg: Uint8Array) => {
          const passphraseEntry = new MockPassphraseEntry();
          const key = new Web3Account(pbkdf2KeyObject, passphraseEntry);
          const sig = await key.sign(msg);
          expect(passphraseEntry.authenticateCalls).toStrictEqual([
            {
              keyId: "3c7bec5e-1f1d-4754-a1ce-3644ce1130f1",
              firstAttempt: true,
            },
          ]);
          expect(passphraseEntry.configurePassphraseCalls).toBe(0);
          return await pubKey.verify(msg, sig);
        }),
      );
    });

    for (const [
      kdf,
      { keyObject, passphrase, privateKey, options },
    ] of Object.entries(testVectors)) {
      const msg = new Uint8Array(128);
      crypto.getRandomValues(msg);

      test(kdf, async () => {
        const passphraseEntry = new MockPassphraseEntry(passphrase);
        const key = new Web3Account(keyObject, passphraseEntry, options);

        expect(passphraseEntry.authenticateCalls).toStrictEqual([]);
        expect(passphraseEntry.configurePassphraseCalls).toBe(0);
        const sig = await key.sign(msg);
        expect(passphraseEntry.authenticateCalls).toStrictEqual([
          { keyId: keyObject.id, firstAttempt: true },
        ]);
        expect(passphraseEntry.configurePassphraseCalls).toBe(0);

        const publicKey = await privateKey.getPublicKey();
        expect(publicKey.verify(msg, sig)).toBeTruthy();

        if (options?.allowWeakPrivateKey) {
          const keyWithStricterOptions = new Web3Account(
            keyObject,
            passphraseEntry,
            {
              allowWeakPrivateKey: false,
            },
          );
          expect(keyWithStricterOptions.sign(msg)).rejects.toThrowError(
            WeakPrivateKeyError,
          );
        }
      });
    }
  });

  describe("exportPrivateKey()", () => {
    for (const [
      kdf,
      { keyObject, passphrase, privateKey, options },
    ] of Object.entries(testVectors)) {
      test(kdf, async () => {
        const passphraseEntry = new MockPassphraseEntry(passphrase);
        const key = new Web3Account(keyObject, passphraseEntry, options);

        expect(passphraseEntry.authenticateCalls).toStrictEqual([]);
        expect(passphraseEntry.configurePassphraseCalls).toBe(0);
        const exported = await key.exportPrivateKey();
        expect(passphraseEntry.authenticateCalls).toStrictEqual([
          { keyId: keyObject.id, firstAttempt: true },
        ]);
        expect(passphraseEntry.configurePassphraseCalls).toBe(0);

        expect(exported).toBeInstanceOf(RawPrivateKey);
        expect(exported.toBytes()).toStrictEqual(privateKey.toBytes());

        if (options?.allowWeakPrivateKey) {
          const keyWithStricterOptions = new Web3Account(
            keyObject,
            passphraseEntry,
            {
              allowWeakPrivateKey: false,
            },
          );
          const exported2 = await keyWithStricterOptions.exportPrivateKey();
          expect(exported2.toBytes()).toStrictEqual(privateKey.toBytes());
        }
      });
    }
  });
});

test("encryptKeyObject()", async () => {
  const privateKey = RawPrivateKey.fromHex(
    "b56fe718d3f29d5f111d3e975abe0ac57595f1717183793540a8065dd5ba34fd",
  );
  await fc.assert(
    fc.asyncProperty(
      fc.uuid(),
      fc.uint8Array(),
      async (keyId: KeyId, passphrase: Passphrase) => {
        const keyObject = await encryptKeyObject(keyId, privateKey, passphrase);
        const decrypted = await decryptKeyObject(keyObject, passphrase);
        expect(decrypted.keyId).toBe(keyId);
        expect(decrypted.privateKey.toBytes()).toStrictEqual(
          privateKey.toBytes(),
        );
      },
    ),
  );
});

describe("decryptKeyObject()", async () => {
  for (const [
    kdf,
    { keyObject, passphrase, privateKey, options },
  ] of Object.entries(testVectors)) {
    test(kdf, async () => {
      const { keyId, privateKey } = await decryptKeyObject(
        keyObject,
        passphrase,
        options,
      );
      expect(keyId).toBe(keyObject.id);
      expect(privateKey.toBytes()).toStrictEqual(privateKey.toBytes());

      await expect(
        decryptKeyObject(keyObject, `wrong ${passphrase}`),
      ).rejects.toThrow(IncorrectPassphraseError);

      if (options?.allowWeakPrivateKey) {
        await expect(
          decryptKeyObject(keyObject, passphrase, {
            allowWeakPrivateKey: false,
          }),
        ).rejects.toThrow(WeakPrivateKeyError);
      }
    });
  }
});
