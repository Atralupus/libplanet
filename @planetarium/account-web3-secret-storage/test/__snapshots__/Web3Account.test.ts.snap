// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Web3Account > exportPrivateKey() 1`] = `
Uint8Array [
  42,
  248,
  218,
  36,
  177,
  119,
  246,
  114,
  61,
  139,
  153,
  118,
  43,
  65,
  111,
  195,
  5,
  2,
  137,
  181,
  114,
  184,
  129,
  158,
  187,
  176,
  166,
  177,
  67,
  177,
  213,
  93,
]
`;
